Add support for adaq776x-1 series and add missing features

jonathanns opened the pull request:
## PR Description

This adds support for ADAQ7768-1, ADAQ7769-1 and ADAQ7767-1 devices from the AD7768-1 family. It also includes:
* Addition of missing IIO attributes
* AAF gain configuration
* PGA control
* Fix on MOSI idle state.
* Support for 16-bit mode, which allows to reach maximum sample rate (1.024 MSPS) with Sinc5 Dec8 filter. 
* Use of new spi-engine implementation.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [x] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -556,35 +791,14 @@ static int ad7768_set_vcm(struct iio_dev *dev,
-	return len;
-}
-
-static IIO_DEV_ATTR_SAMP_FREQ_AVAIL(ad7768_sampling_freq_avail);
```
This is a no-go... This breaks ABI and will never be accepted upstream. Why are we removing this attr?

jonathanns replied:
I think because we create an attribute to each configuration (Filter, decimation and mclk divider ), so the sampling frequency would be a result of the combination of them.

I can remove one of these (like the mclk divider you mentioned above) and keep this attr.

nunojsa replied:
> I think because we create an attribute to each configuration (Filter, decimation and mclk divider ), so the sampling frequency would be a result of the combination of them.

Not sure if possible at all or even remotely doable but the natural way of doing things would be to keep sampling frequency and try to get a combination of Filter, decimation and mclk divider to get the desired frequency (when someone sets that attr). Or maybe trying to somehow "squash" mclk and decimation in the sampling_freq and having the filter as a separate control. You still need to take care because changing the filter will likely affect the sampling freq...

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -494,46 +562,213 @@ int ad7768_gpio_init(struct ad7768_state *st)
+
+	/* A sync-in pulse is required every time the filter dec rate changes */
+	gpiod_set_value(st->gpio_sync_in, 1);
+	gpiod_set_value(st->gpio_sync_in, 0);
```
I see now this is moved? Do we really have to move it?

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -157,13 +225,17 @@ static const char * const ad7768_vcm_modes[] = {
 
 static const char * const ad7768_filter_enum[] = {
 	[SINC5] = "SINC5",
-	[SINC5_DEC_X8] = "SINC5_DEC_X8",
```
so we added this attr in a previous patch and we are now removing it? That's nok and seems to me that your patches need to be re-arranged somehow

jonathanns replied:
indeed, i will work on that.

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -235,6 +305,13 @@ static const struct iio_enum ad7768_mclk_div_iio_enum = {
+	.items = ad7768_filter_low_freq_enum,
+	.num_items = ARRAY_SIZE(ad7768_filter_low_freq_enum),
+	.get = ad7768_get_dig_fil,
+};
```
Do we really need a dedicated attr for this?

jonathanns replied:
So we shouldn't have this attribute when the 16-bit mode is activated?

nunojsa replied:
What is this attr about? Has it something to do with?

[IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY](https://elixir.bootlin.com/linux/v6.11/C/ident/IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY)

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -270,28 +347,76 @@ static struct iio_chan_spec_ext_info ad7768_ext_info[] = {
-			.shift = 0,
-		},
-	},
+	AD7768_CHAN(0, AD7768_CHAN_INFO_NONE, 24, ad7768_ext_info),
```
This should be in a different patch... Like a patch in preparation for adding support for more devices.

jonathanns replied:
Right!

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -270,28 +347,76 @@ static struct iio_chan_spec_ext_info ad7768_ext_info[] = {
+	const struct iio_chan_spec *channel_spec[2];
+	const unsigned long *available_masks;
+	int num_channels;
+};
```
you could also introduce the chip_info struct in a preparatory patch. The goal is to make patches smaller and easier to review

jonathanns replied:
Noted

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -791,12 +935,64 @@ static int ad7768_set_vcm(struct iio_dev *dev,
+				   AD7768_GPIO_WRITE(pgia_pins_value));
+	if (ret < 0)
+		return ret;
+	st->pga_gain_mode = gain_mode;
```
This function seems to need proper locking

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -898,14 +1141,19 @@ static int ad7768_setup(struct ad7768_state *st)
 
 	st->gpio_sync_in = devm_gpiod_get(&st->spi->dev, "adi,sync-in",
-					  GPIOD_OUT_LOW);
+					  GPIOD_IN);
```
Why? Your commit messages need more details :). Also note this is also changing behavior for other devices so this needs to be explained

jonathanns replied:
Right, I will check this again. 
We changed to input because we are no longer using a GPIO to drive the SYNC_IN. Instead, we are sending a sync command to the device and connecting the sync-out pin to the sync-in. If we kept it as an output, it would conflict with the sync-out output.

I can also create a different patch.

nunojsa replied:
Hmm, it actually seems that this is already a GPO in the uptream version of this driver

jonathanns replied:
I see. What if I remove the `adi,sync-in-gpios`  from the required properties and if it is not present, I use the SPI SYNC instead? Of course that would require the SYNC_IN to be connected to SYNC_OUT in the board.

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1082,10 +1355,59 @@ static int ad7768_set_channel_label(struct iio_dev *indio_dev,
+		.has_pga = true,
+		.has_variable_aaf = true
+	}
+};
```
use a different struct per device and drop the array please

jonathanns replied:
Ok, noted.

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -947,15 +1195,12 @@ static irqreturn_t ad7768_interrupt(int irq, void *dev_id)
 	int ret;
 
+	st->offload_xfer.len = roundup_pow_of_two(BITS_TO_BYTES(st->bits_per_word));
+	st->offload_xfer.bits_per_word = st->bits_per_word;
```
This looks a bit suspicious... Isn't this already about supporting the new offload engine implementation?

jonathanns replied:
yes, that could be a preparation as well, but it is more related to the other patch, so i will move it there.

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1116,21 +1438,63 @@ static int ad7768_probe(struct spi_device *spi)
+		if (!st->chip)
+			return dev_err_probe(&spi->dev, -ENODEV,
+					     "Could not find chip info data\n");
+	}
```
you can the above in one call...

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1116,21 +1438,63 @@ static int ad7768_probe(struct spi_device *spi)
-	indio_dev->name = spi_get_device_id(spi)->name;
+	ret = device_property_read_u32(&spi->dev, "adi,low-latency", &val);
+	if (ret)
+		return ret;
```
I don't think this needs to be a mandatory property at all... Also seems like a boolean one. But more importantly, it raises the question about this being a DT property? It's at the very least not too coherent with the power_modes attr? Is this somehow related with those modes?

jonathanns replied:
Yes, that could be a boolean. We need something like this in the devicetree to set the IIO channel to either 16 or 24 bits. 

In Sinc5 Decimation x8 mode, the sample width is set to 16 bits, while in other filter modes, the sample width is 24 bits.. Since we cannot change the IIO channel in runtime, we need a property in the device tree to choose between the two cases.

When this property is "enabled," the device will operate exclusively in Sinc5 Dec8 mode (16-bit samples). If the property is disabled, the filter mode can be chosen between the other options, and the sample width will default to 24 bits.

nunojsa replied:
> In Sinc5 Decimation x8 mode, the sample width is set to 16 bits, while in other filter modes, the sample width is 24 bits.. Since we cannot change the IIO channel in runtime, we need a property in the device tree to choose between the two cases.

If I'm understanding correctly, you can now change the scan elements at run time... take a look at 

https://elixir.bootlin.com/linux/v6.11/source/include/linux/iio/iio.h#L839

and how it's being used...

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1116,21 +1438,63 @@ static int ad7768_probe(struct spi_device *spi)
+	st->aaf_gain = AD7768_AAF_IN1;
+
+	if (device_property_present(&spi->dev, "adi,aaf-gain")
+	    && st->chip->has_variable_aaf) {
```
Don't check for presence... Instead, do it like this:

```
ret = device_property_read_u32(&spi->dev, "adi,aaf-gain", &val);
if (!ret) {
    if (!st->chip->has_variable_aaf)
        return dev_err_probe(...);

   // proceed with code validating the prop value
   ...
}
```

jonathanns replied:
noted

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1415,6 +1415,15 @@ static int ad7768_probe(struct spi_device *spi)
+	* Request the SPI controller to make MOSI idle high.
+	*/
+	spi->mode |= SPI_MOSI_IDLE_HIGH;
+	ret = spi_setup(spi);
```
This should have a Fixes tag and be the first patch in the pull. Fixes should always come first in case we want to backport them.

Also wondering why this was never caught...

jonathanns replied:
Ok, i will separate the fixes and put at the beginning

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -4,6 +4,7 @@
  *
  * Copyright 2017 Analog Devices Inc.
  */
+#include "linux/byteorder/little_endian.h"
```
this seems like an header being auto added by clangd :sweat_smile:

jonathanns replied:
Indeed, i will check this out

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -458,16 +459,16 @@ static int ad7768_spi_reg_read(struct ad7768_state *st, unsigned int addr,
 	if (ret < 0)
 		return ret;
-	*data = (len == 1 ? st->data.buf[0] : st->data.word);
+	*data = (len == 1 ? st->data.buf[1] : cpu_to_be32(st->data.word));
```
are the above changes needed?

jonathanns replied:
It is necessary for the new spi-engine integration

jonathanns commented:
> First round of review... There#s some stuff with ABI that needs to be figured. In your second patch, you also have:
> 
> "It also fixes sampling frequency calculation, sync pulse using registers and data acquisition with iio."
> 
> Please don't mix fixes with new code being added. Fixes need to come first and be as standalone as possible.
> 
> Also note that I'm aware this driver already diverged from the upstream version but it would very nice (and it would make a lot of sense) if you could send these patches upstream with the new devices and new ABI. Naturally you would have to ommit the spi offload part of it but the new ABI would be very important to have it discussed (and agreed) upstream. Otherwise, what it will most likely happen is that we merge something in here that will then change when upstreaming the missing bits of this driver upstream.

Thanks for the feedback, I will re-arrange the commits and take a look at the ABI issue. I am sending these patches now to get a preliminar review before sending them upstream. So, when it is good enough, we can hold off this PR until they are upstreamed.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -494,46 +562,213 @@ int ad7768_gpio_init(struct ad7768_state *st)
+		return ret;
+	mclk_div_value = (val & 0xcf) | AD7768_PWR_MCLK_DIV(mclk_div);
+
+	ret = ad7768_spi_reg_write(st, AD7768_REG_POWER_CLOCK, mclk_div_value);
```
ok

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -494,46 +562,213 @@ int ad7768_gpio_init(struct ad7768_state *st)
+	if (ret)
+		return ret;
+	mode = (val & 0xf8) | AD7768_DIG_FIL_DEC_RATE(dec_rate);
+	ret = ad7768_spi_reg_write(st, AD7768_REG_DIGITAL_FILTER, mode);
```
ok

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -494,46 +562,213 @@ int ad7768_gpio_init(struct ad7768_state *st)
+	if (ret)
+		return ret;
+	mode = (val & 0xfc) | AD7768_PWR_PWRMODE(pwr_mode);
+	ret = ad7768_spi_reg_write(st, AD7768_REG_POWER_CLOCK, mode);
```
right

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -494,46 +562,213 @@ int ad7768_gpio_init(struct ad7768_state *st)
+			goto out;
+		readin = clamp_t(long long, readin, 0, 8191);
+		st->sinc3_dec_rate = readin;
+		dec_rate_msb = (readin & 0x1f00) >> 8;
```
noted

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -309,13 +399,13 @@ static int ad7768_scan_direct(struct iio_dev *indio_dev)
 		reinit_completion(&st->completion);
 
-	ret = ad7768_set_mode(st, AD7768_ONE_SHOT);
+	ret = ad7768_set_mode(st, AD7768_CONTINUOUS);
```
Yes, i will have this separate as well

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -174,12 +250,23 @@ static const struct iio_enum ad7768_vcm_mode_enum = {
+	IIO_ENUM("dec_rate", IIO_SHARED_BY_ALL, &ad7768_dec_rate_iio_enum),
+	IIO_ENUM_AVAILABLE("dec_rate", IIO_SHARED_BY_ALL, &ad7768_dec_rate_iio_enum),
+	IIO_ENUM("mclk_div", IIO_SHARED_BY_ALL, &ad7768_mclk_div_iio_enum),
+	IIO_ENUM_AVAILABLE("mclk_div", IIO_SHARED_BY_ALL, &ad7768_mclk_div_iio_enum),
```
MCLK is an input clock generated by a Crystal. `mclk_div ` is set by a configuration register to control the sampling frequency.

nunojsa replied:
Could we then somehow implement this as part of the sampling_frequency attr?

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -150,21 +155,92 @@ static const char * const ad7768_vcm_modes[] = {
+	[AD7768_ECO_MODE]   = "AD7768_ECO_MODE",
+	[AD7768_MED_MODE]   = "AD7768_MED_MODE",
+	[AD7768_FAST_MODE]  = "AD7768_FAST_MODE",
+};
```
I will take a look at the Power management module.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -150,21 +155,92 @@ static const char * const ad7768_vcm_modes[] = {
-	{ AD7768_MCLK_DIV_4, AD7768_DEC_RATE_512, 2048, AD7768_MED_MODE },
-	{ AD7768_MCLK_DIV_4, AD7768_DEC_RATE_1024, 4096, AD7768_MED_MODE },
-	{ AD7768_MCLK_DIV_8, AD7768_DEC_RATE_1024, 8192, AD7768_MED_MODE },
-	{ AD7768_MCLK_DIV_16, AD7768_DEC_RATE_1024, 16384, AD7768_ECO_MODE },
```
I guess we divided it into separete attributes

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -174,12 +250,23 @@ static const struct iio_enum ad7768_vcm_mode_enum = {
+	IIO_ENUM("common_mode_voltage", IIO_SHARED_BY_ALL, &ad7768_vcm_mode_enum),
+	IIO_ENUM_AVAILABLE("common_mode_voltage", IIO_SHARED_BY_ALL, &ad7768_vcm_mode_enum),
+	{
+		.name = "sinc3_dec_rate",
```
This case is a bit more complicated, the device has two configurable decimation rates: one for wideband and sinc5 filter  modes (with 6 decimation options) and other for sinc3 filter mode (with 1024 possible values). Thats why we have two decimation attributes.

Other issue is that we have other two filter modes, `sinc5 dec8 `and `sinc5 dec16`, where these attributes don't matter, the decimation is fixed.

nunojsa replied:
Yes, but my point is to see if we can extend the existing ABI with the new modes supported by this device other than adding new ABI.

jonathanns commented:
v2:
* Re-arranged commits to prioritize fixes and include preparatory patches.
* Ported [Multiple scan type feature ](https://lore.kernel.org/all/20240530-iio-add-support-for-multiple-scan-types-v3-1-cbc4acea2cfa@baylibre.com/t/#u) from upstream.
* Added support for the Multiple Scan Type feature, allowing IIO context changes at runtime. This replaces the low-latency property in the device tree.
* Restored the Sampling Frequency Available attribute. While it still controls decimation and the MCLK divider, it no longer affects filter mode, which is now managed by a separate attribute. Consequently, the decimation and MCLK divider attributes have been removed.
* The Sampling Frequency Available attribute now shows a range of values instead of a list, to include all supported sampling frequencies, including those for the Sinc3 filter.
* Filter mode attribute more aligned to AD4130's ABI.
* Dropped unrelated and unecessary changes.

machschmitt commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -881,11 +1052,16 @@ static int ad7768_read_raw(struct iio_dev *indio_dev,
+			mutex_unlock(&st->lock);
+			return IIO_VAL_INT_PLUS_NANO;
+		}
+		*val = (st->vref * 2) / 1000;
```
same here, probably safer to subtract the precision bit by 1 if the channel is true bipolar.

machschmitt commented on arch/arm/boot/dts/zynq-zed-adv7511-ad7768-1-evb.dts:
```diff
@@ -83,7 +83,7 @@
 			spi-cpha;
 			vref-supply = <&vref>;
-			adi,sync-in-gpios = <&gpio0 88 GPIO_ACTIVE_LOW>;
+			adi,sync-in-gpios = <&gpio0 92 GPIO_ACTIVE_LOW>;
```
Why do you need to update this?
Didn't you say sync-out was connected to sync-in?

machschmitt commented on arch/arm/boot/dts/zynq-zed-adv7511-adaq7767-1-evb.dts:
```diff
@@ -0,0 +1,96 @@
+			spi-cpol;
+			spi-cpha;
+			vref-supply = <&vref>;
+			adi,sync-in-gpios = <&gpio0 92 GPIO_ACTIVE_LOW>;
```
Will the sync-in GPIO be needed if SYNC_OUT is connected to SYNC_IN?

machschmitt commented on arch/arm/boot/dts/zynq-zed-adv7511-adaq7768-1-evb.dts:
```diff
@@ -0,0 +1,95 @@
+			spi-cpol;
+			spi-cpha;
+			vref-supply = <&vref>;
+			adi,sync-in-gpios = <&gpio0 92 GPIO_ACTIVE_LOW>;
```
same question here, sync-in needed?

machschmitt commented on arch/arm/boot/dts/zynq-zed-adv7511-adaq7769-1-evb.dts:
```diff
@@ -0,0 +1,96 @@
+			spi-cpol;
+			spi-cpha;
+			vref-supply = <&vref>;
+			adi,sync-in-gpios = <&gpio0 92 GPIO_ACTIVE_LOW>;
```
sync-in needed?

machschmitt commented on Documentation/devicetree/bindings/iio/adc/adi,ad7768-1.yaml:
```diff
@@ -39,6 +48,16 @@ properties:
+
   adi,sync-in-gpios:
     maxItems: 1
     description:
```
good to also have a dt check for the gain property.
```
allOf:
  # Gain property only applies to ADAQ devices
  - if:
      properties:
        compatible:
          not:
            contains:
              enum:
                - adi,adaq7767-1
                - adi,adaq7768-1
                - adi,adaq7769-1
    then:
      properties:
        adi,gain-milli: false
```

machschmitt reviewed (commented):
This PR adds support for a lot of stuff.
In some patch I saw something like `Co-authored-by` when the official tag is `Co-developed-by`. See Documentation/process/submitting-patches.rst.
The dt-bindings patch should come before the updates to the dts files.
dts files should follow the documentation provided by the dt (binding) doc.
Patches can probably become more concise and cleaner.
Anyways, looks like you're on the right track for it.

machschmitt commented:
Also, FWIW, the updates to adaq7768-1 driver use the new MOSI idle feature which is not included in this PR but is included in https://github.com/analogdevicesinc/linux/pull/2621. So that's probably the cause of some of the CI checks never passing here.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -286,6 +368,25 @@ static int ad7768_spi_reg_write_masked(struct ad7768_state *st,
+	unsigned int val, mclk_div_value;
+	int ret;
+
+	mutex_unlock(&st->lock);
```
My mistake. I will fix that

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -825,6 +843,11 @@ static int ad7768_read_raw(struct iio_dev *indio_dev,
 {
 	struct ad7768_state *st = iio_priv(indio_dev);
 	int scale_uv, ret;
+	const struct iio_scan_type *scan_type;
```
noted

machschmitt commented:
Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________
From: Jonathan Santos ***@***.***>
Sent: Monday, October 21, 2024 12:40:54 PM
To: analogdevicesinc/linux ***@***.***>
Cc: Schmitt, Marcelo ***@***.***>; Review requested ***@***.***>
Subject: Re: [analogdevicesinc/linux] Add support for adaq776x-1 series and add missing features (PR #2587)

[External]


@jonathanns commented on this pull request.

________________________________

In drivers/iio/adc/ad7768-1.c<https://urldefense.com/v3/__https://github.com/analogdevicesinc/linux/pull/2587*discussion_r1809063399__;Iw!!A3Ni8CS0y2Y!_gq45BQs1W4FdBJIyq6hlxteUGYfF3_B7SOw931waHjWBSgxpYR3SsAkJqR6ow7DhpNRHK-nuOSGyvVHylUE8I_vUnB2$>:

> @@ -898,14 +1141,19 @@ static int ad7768_setup(struct ad7768_state *st)
                return ret;

        st->gpio_sync_in = devm_gpiod_get(&st->spi->dev, "adi,sync-in",
-                                         GPIOD_OUT_LOW);
+                                         GPIOD_IN);


I see. What if I remove the adi,sync-in-gpios from the required properties and if it is not present, I use the SPI SYNC instead? Of course that would require the SYNC_IN to be connected to SYNC_OUT in the board.

The dt-binding can have a allOf/oneOf clause. Something like:

required:

  allOf:

    - reg

    - other required props ...

    - oneOf:

      - adi,sync-in-gpios

      - adi,sync-out-to-sync-in

correct syntax may be different from the above.

Additional checks may be included if needed or wanted. Example:

allOf:

  - if

    adi,sync-in-gpios;

  - then

    adi,sync-out-to-sync-in: false;

Have a look at other bindings to get the correct syntax for those checks.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/analogdevicesinc/linux/pull/2587*discussion_r1809063399__;Iw!!A3Ni8CS0y2Y!_gq45BQs1W4FdBJIyq6hlxteUGYfF3_B7SOw931waHjWBSgxpYR3SsAkJqR6ow7DhpNRHK-nuOSGyvVHylUE8I_vUnB2$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/A5OAY4CBDX3CQKMIA6VGYIDZ4UOABAVCNFSM6AAAAABNHCV5FWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGOBSGUYTEMBUGA__;!!A3Ni8CS0y2Y!_gq45BQs1W4FdBJIyq6hlxteUGYfF3_B7SOw931waHjWBSgxpYR3SsAkJqR6ow7DhpNRHK-nuOSGyvVHylUE8McodytN$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

jonathanns commented:
v3:

- Added a new `adi,sync-in-spi` spi property. Now the user can choose between GPIO and SPI synchronization. The option to omit both is also supported for cases with multiple devices, where the synchronization signal comes from a external source, such as a "master" device.
- Introduced additional checks in the DT bindings.
- Removed changes to the `sync-in` gpio number. Turns out that there was no reason for that and this has been reverted in the HDL.
- Addressed some code style issues.
- Removed the multiplaction of vref by 2 and reduced the precision by one. But still, how do I check if the input is Bipolar or not? From what i've  seen in similar devices, we just assume the Bipolar case, unless the ADC have some configuration register.
- Replaced `Co-authored-by` for `Co-developed-by`

machschmitt commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -343,7 +345,7 @@ static int ad7768_spi_reg_read(struct ad7768_state *st, unsigned int addr,
 	unsigned char tx_data[4];
 	int ret;
 
-	tx_data[0] = AD7768_RD_FLAG_MSK(addr);
+	tx_data[len] = AD7768_RD_FLAG_MSK(addr);
```
Undo this too when proper endianness handling is set in scan_type struct.

machschmitt commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1276,6 +1525,7 @@ static int ad7768_probe(struct spi_device *spi)
 	st->mclk_freq = clk_get_rate(st->mclk);
 	st->spi_is_dma_mapped = spi_engine_ex_offload_supported(spi);
 	st->irq = spi->irq;
+	st->vref = regulator_get_voltage(st->regulator);
```
looks like `st->regulator` is only to get the ADC reference voltage (vref). If that's the case, you can simplify this with `devm_regulator_get_enable_read_voltage()` now in ADI main.

jonathanns replied:
Right!

machschmitt commented on Documentation/devicetree/bindings/iio/adc/adi,ad7768-1.yaml:
```diff
@@ -65,7 +106,21 @@ required:
   - vref-supply
   - spi-cpol
   - spi-cpha
-  - adi,sync-in-gpios
```
don't do it all together. One patch for the new parts, another one for the updates to sync pins.
Also, the dt-binding patches should come before the driver changes that use the new dt properties.

machschmitt replied:
put the gain property and related checks in the same patch that adds the new parts and put the changes related to `adi,sync-in-spi` property in a separate patch. That's what I wanted to say previously.

machschmitt commented:
Hi @jonathanns, the series is looking better.  A few more comments:

The patches for multiple scan_types have been ported to ADI main. Rebase on top of main so you won't need to carry them in this PR anymore.

Tip: try to add at least one imperative sentence to concisely state what patches do.
For example, the "fix MOSI idle state" patch looks good, but it's log/body doesn't really say what it's doing.
Add something like "Set SPI MOSI idle high configuration to keep the data line high when idle." to the end of the description to make it much more clear.


>     * Removed the multiplaction of vref by 2 and reduced the precision by one. But still, how do I check if the input is Bipolar or not? From what i've  seen in similar devices, we just assume the Bipolar case, unless the ADC have some configuration register.

ADC input type naming is confusing. I've been trying to clarify the relation between ADC input configuration and IIO support for them in this candidate doc https://github.com/analogdevicesinc/linux/pull/2573/commits/1373835f9e16ce6ce3f5072ca6026f37461eb84f. I still need to verify the how accurate (or not) those concepts are with other ADI engineers though. The goal is to have something to help us be less uncertain about IIO channel configuration in the future. Answering your question, an ADC channel is considered bipolar if its configuration allows it to output data that represents a negative quantity. For example, if a differential ADC has a pair of inputs (IN+ and IN-) working as a differential input, then a negative value must be provided to user space when IN+ < IN- to indicate that IN+ is below/smaller than IN-. Not that can happen when inputs are above or even below GND. When inputs can go below GND they are called true bipolar but the most important characteristic is whether the input configuration allows/requires negative values to be expressed.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -373,23 +474,92 @@ static int ad7768_reg_access(struct iio_dev *indio_dev,
+	st->filter_mode = filter_mode;
+	ret = ad7768_spi_reg_read(st, AD7768_REG_DIGITAL_FILTER, &mode, 1);
+	if (ret)
+		goto out;
```
Sure

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1029,15 +1062,20 @@ static irqreturn_t ad7768_interrupt(int irq, void *dev_id)
-		.bits_per_word = 32
 	};
 	unsigned int rx_data[2];
-	unsigned int tx_data[2];
```
The device has a continuos read mode that makes unecessary to give the 8 bits address to read the ADC sample. 
This is related to the precision mode because we cannot reach1.024 MSPS at a 20MHz SPI clock if the transfer is bigger than 16 bits.

- with the address write: (16 + 8) * 1.024 MHz > 20MHz
- without the address write: 16 * 1.024 MHz < 20MHz


We only reduce the precision from 24 to 16 bits when we set the sinc5-dec8 filter (this mode enables configuring the ODR to 1.024 MHz).

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1050,8 +1088,6 @@ static int ad7768_buffer_postenable(struct iio_dev *indio_dev)
 	if (st->spi_is_dma_mapped) {
 		spi_bus_lock(st->spi->master);
 
-		tx_data[0] = AD7768_RD_FLAG_MSK(AD7768_REG_ADC_DATA) << 24;
-		xfer.tx_buf = tx_data;
```
Both cases actually don't require the address byte if continuous read mode is enabled. However, for 16-bit mode to work correctly, it’s mandatory to remove the address byte, so we had to adjust for that.

There is already a comment in this same function regarding the continuos read mode, so I don't know why we were using the `tx_data` before
```

       /*
	* Write a 1 to the LSB of the INTERFACE_FORMAT register to enter
	* continuous read mode. Subsequent data reads do not require an
	* initial 8-bit write to query the ADC_DATA register.
	*/
	ret =  ad7768_spi_reg_write(st, AD7768_REG_INTERFACE_FORMAT, 0x01);

```

jonathanns commented:
v4:

- Splitted the dt-binding patch: one for the new supported parts and other for new properties and checks.
- moved dt-related patches to the start of the tree.
- changed the commit messages to include at least one imperative sentence.
- Moved the removal of tx_buf from spi-engine transfers in the offload to a separate patch.
- Added Decimation rate attribute: preivous sampling frequency attribute obscured the trade-offs by forcing a combination of mclk_div and decimation rate.
- Sampling frequency now calculates only the mclk_div value.
- Power mode attribute removed. this value is set based on the value of mclk_div (as the datasheet recommends).
- Changed spi register write and read to fix endianness problem.

machschmitt reviewed (commented):
Looking much better this version.

Though, there is still room for improvement.
minor commit log neat:
would avoid the word "fix" and try to be more informative when possible. e.g.
"iio: adc: ad7768-1: fix MOSI idle state" -> "iio: adc: ad7768-1: Set MOSI idle state to high"

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -371,23 +454,136 @@ static int ad7768_reg_access(struct iio_dev *indio_dev,
+	unsigned int mode, dec_rate_reg;
+	int ret;
+
+	mutex_lock(&st->lock);
+
+	dec_rate_reg = find_closest(dec_rate, dec_rate_values,
+				    ARRAY_SIZE(dec_rate_values));
```
Right, sure.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -574,13 +842,16 @@ static ssize_t ad7768_sampling_freq_avail(struct device *dev,
+					 ad7768_mclk_div_rates[i]);
+		/* Sampling frequency cannot be lower than the minimum of 50 SPS */
+		if (freq >= 50)
+			len += scnprintf(buf + len, PAGE_SIZE - len, "%d ", freq);
```
Noted

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -590,6 +861,38 @@ static ssize_t ad7768_sampling_freq_avail(struct device *dev,
+	/* Return decimation rate available in range format */
+	buf[len++] = '[';
+	if (st->filter_mode == SINC3) {
+		len += scnprintf(buf + len, PAGE_SIZE - len, "%d ", SINC3_DEC_RATE_MIN);
```
I believe `sysfs_emit_at` fits better, but thanks!

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1077,15 +1136,21 @@ static irqreturn_t ad7768_interrupt(int irq, void *dev_id)
+	const struct iio_scan_type *scan_type;
 	struct spi_transfer xfer = {
 		.len = 1,
-		.bits_per_word = 32
```
I don't know... I am removing this because I need to assign `bits_per_word `  with the  `scan_type->realbits`, as below:


```
xfer.bits_per_word = scan_type->realbits;
```

machschmitt replied:
makes sense. That's good then :)

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -421,6 +458,9 @@ static int ad7768_scan_direct(struct iio_dev *indio_dev)
 	ret = ad7768_spi_reg_read(st, AD7768_REG_ADC_DATA, &readval, 3);
 	if (ret < 0)
 		return ret;
+
+	if (st->filter_mode == SINC5_DEC_X8)
+		readval = readval >> 8;
```
If i do this in a previous patch, it will be inconsistent with the scale value. This patch adds the 16-bit resolution, so anything before is just 24-bit.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -272,23 +274,34 @@ static struct iio_chan_spec_ext_info ad7768_ext_info[] = {
+		.ext_info = ad7768_ext_info,\
+		.indexed = 1,\
+		.channel = _idx,\
+		.scan_index = 0,\
```
noted

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1279,6 +1292,18 @@ static int ad7768_set_channel_label(struct iio_dev *indio_dev,
+static const struct ad7768_chip_info ad7768_chip_info = {
+	.name = "ad7768-1",
+	.channel_spec = ad7768_channels,
+	.num_channels = 1,
```
Right!

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -995,15 +1166,53 @@ static int ad7768_read_raw(struct iio_dev *indio_dev,
 		return ad7768_set_freq(st, val);
+	case IIO_CHAN_INFO_SCALE:
+		if (!st->chip->has_pga)
+			return -EPERM;
```
ok

machschmitt replied:
actually, I think `-EOPNOTSUPP` would be more appropriate error code here.

jonathanns replied:
Right!

jonathanns commented:
v5:

- Improved some commit messages.
- Added channel subnode to new devicetrees as recommended.
- Made SPI and GPIO Synchronization mutually exclusive in the driver.
- Fixed multiline comments.
- Fixed error handling in some functions.
- Fixed readability issues.
- Replaced `sprintf `and `scnprintf` by `sysfs_emit()` and `sysfs_emit_at()`.
- Use of `ARRAY_SIZE()` to represent number of channels.
- Added Scan type signal check to fix scale computation.
- Replaced `vref `by `vref_uv` so the voltage unit is more clear.
- Reorganized Documentation patches as requested: First new supported parts and aff gain, followed by `adi,sync-in-spi` support.

machschmitt commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -652,6 +769,62 @@ static int ad7768_get_dig_fil_attr(struct iio_dev *dev,
+			goto out;
+	}
+
+	/* Write GPIOs 0-2 with the gain mode equivalente value */
```
minor neat: write English comments
"Write GPIOs 0-2 with the gain mode equivalente value" -> "Write the respective gain values to GPIOs 0, 1, 2"

machschmitt reviewed (commented):
A few more comments. Most of them would be minor changes to the driver.

jonathanns commented on arch/arm/boot/dts/zynq-zed-adv7511-adaq7767-1-evb.dts:
```diff
@@ -0,0 +1,103 @@
+			spi-cpha;
+			vref-supply = <&vref>;
+			adi,sync-in-spi;
+			adi,aaf-gain = <143>;
```
Sorry, At some point I changed from `adi,gain-milli` to `adi,aaf-gain`. Is the first option more adequate? I can replace it

machschmitt replied:
It depends on how the gain is applied. `adi,gain-milli` is used to describe the **hardware gain** applied to the ADC inputs. For ad4000 series, the gain provided by the ADC input scaler is **defined by the hardware connections** between chip pins OUT+, R1K-, R1K1-, R1K+, R1K1+, and OUT-. For gains set through register configuration or runtime (software) settings, IIO interfaces should be used (_scale, _callibscale). What is this AAF about? Is it related to datasheet AFE_GAIN?

jonathanns replied:
AFE_gain is the combination of Anti-Aliasing Filter (AAF) gain and PGA gain.  AAF is an analogic filter with three inputs, each one lead to a specific gain and bandwidth rejection. The user must connect the output of the first filter (PGA), for the case of adaq7769-1,  to one of these AAF inputs by inserting a resistor in the corresponding pin:

![image](https://github.com/user-attachments/assets/5c4111f3-9e85-4eab-91c0-640617aec90f)

Only the ADAQ7769-1 and ADAQ7767-1 have variable AAF gain.  The second one does not have PGIA, so the AAF input is directly connected to the signal input and AFE gain is the AAF gain.

machschmitt replied:
Hmm, that's a bit different than `adi,gain-milli` in that it also defines a bandwidth rejection. Though, to our luck, there is no IIO ABI for exposing bandwidth rejection settings to user space so I think it's okay to reuse `adi,gain-milli` to describe the AFF gain. Alternatively, if you want, you can instead provide a property with a different name (e.g. `adi,aaf-gain`) documenting it specifies both input gain and bandwidth rejection. In either case, include the documentation for it in Documentation/devicetree/bindings/iio/adc/adi,ad7768-1.yaml.

machschmitt replied:
About the device that doesn't have variable AAF gain, I think it may be nice to add a check for it. e.g.
```
  # Variable AAF only applicable to ADAQ7769-1 and  ADAQ7767-1
  - if:
      properties:
        compatible:
          not:
            contains:
              enum:
                - adi,adaq7767-1
                - adi,adaq7769-1
    then:
      properties:
        adi,aaf-gain: false

```

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -506,44 +713,123 @@ int ad7768_gpio_init(struct ad7768_state *st)
+	long long val;
+	struct ad7768_state *st = iio_priv(indio_dev);
+
+	mutex_lock(&st->lock);
```
ok!

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -157,6 +157,19 @@ enum ad7768_dec_rate {
+	enum ad7768_mclk_div mclk_div;
+	enum ad7768_dec_rate dec_rate;
+	unsigned int clk_div;
+};
```
This must be a leftover from the time I remade all commits. I will remove it, thanks

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1297,6 +1310,18 @@ static int ad7768_set_channel_label(struct iio_dev *indio_dev,
+	.name = "ad7768-1",
+	.channel_spec = ad7768_channels,
+	.num_channels = ARRAY_SIZE(ad7768_channels),
+	.available_masks = ad7768_channel_masks,
```
Rigth, I am going to add it

jonathanns commented:
v6:
* Replaced `adi-gain-milli` for `adi-aaf-gain` on dts binding.
* Fixed dts bindings check - grouped them into the same allOf key.
* Added read of `available_masks `on probe.
* Removed copyrigth line from ad7768-1 devicetree
* Removed unsed `ad7768_clk_configuration ` struct.
* Addressed other nits.

machschmitt commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -1168,14 +1167,19 @@ static int ad7768_buffer_postenable(struct iio_dev *indio_dev)
 		return ret;
 
 	if (st->spi_is_dma_mapped) {
+		st->offload_xfer.rx_buf = rx_data;
```
Don't know the detailed explanation for it but I've been told to set buffers with a hacky pointer when the data is expected to travel through DMA. I suppose that would apply to the offload buffer here as well.
```c
	/* HACK: invalid pointer indicates read data from SPI offload DMA */
	st->offload_xfer.rx_buf = (void*)(-1);
```

machschmitt commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -985,12 +1158,15 @@ static int ad7768_read_raw(struct iio_dev *indio_dev,
+		*val = st->vref_uv / 1000;
+		if (st->chip->has_variable_aaf)
+			*val = (*val * MILLI) / ad7768_aaf_gains[st->aaf_gain];
+		*val2 = scan_type->realbits - 1;
```
Was about to suggest something like
```c
if (scan_type->sign == 's')
	*val2 = scan_type->realbits - 1;
else
	*val2 = scan_type->realbits;
```
but the ADC output code is always in two's complement format so the current `*val2` assignment is correct.
If you think appropriate, add a comment like
```
/* ADC output code is two's complement so only (realbits - 1) bits express voltage magnitude */
```

machschmitt reviewed (commented):
Looks good overall.
It will certainly change much more if submitted upstream but, for now, I think this is fairly good. With the last of my comments addressed,
Reviewed-by: Marcelo Schmitt <marcelo.schmitt@analog.com>

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -981,11 +1020,25 @@ static const struct attribute_group ad7768_group = {
+	if (st->spi_is_dma_mapped)
+		return st->filter_mode == SINC5_DEC_X8 ? AD7768_SCAN_TYPE_DMA_HIGH_SPEED :
+		       AD7768_SCAN_TYPE_DMA_NORMAL;
+	else
```
redundant `else`

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -180,6 +187,33 @@ static const  int sinc3_dec_rate_max_values[4] = {
+		.storagebits = 32,
+		.endianness = IIO_CPU,
+	},
+};
```
Just one thing that I don't understand... The filter mode was implemented in the previous patch right? So if I'm not missing nothing it seems the previous patch has a bug for some filter or decimation values? If I'm right, this should be a precursor patch...

nunojsa commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -568,6 +642,42 @@ static int ad7768_set_dec_rate(struct ad7768_state *st, unsigned int dec_rate)
+		if (st->chip->has_variable_aaf) {
+			numerator *= ad7768_aaf_gains[st->aaf_gain];
+			denominator *= MILLI;
+		}
```
nit: newline

nunojsa reviewed (commented):
Just some comments on top of @marceloschmitt1. Please go ahead and usptream the bits that can be upstreamed.

jonathanns commented on Documentation/devicetree/bindings/iio/adc/adi,ad7768-1.yaml:
```diff
@@ -127,6 +134,20 @@ allOf:
+        adi,sync-in-spi: false
+  - if:
+      required:
+        - adi,sync-in-spi
```
Here's what we have decided for the synchronization:

- Using GPIO: We define `adi,sync-in-gpios` and route the GPIO pin to the SYNC_IN pin on the board.
- Using SPI command: We define the `adi,sync-in-spi` and route the SYNC_OUT pin to the SYNC_IN pin in the eval board.
- We also have the option to do use none of them, but in this case the SYNC_IN pulse should be providaded by an external device or another source (e.g., when multiple devices are involved).

If you believe the third option doesn’t need support, we could consider the absence of adi,sync-in-gpios as indicative of the SPI command synchronization. I think it’s good to include the adi,sync-in-spi option, because it clarifies the need to route SYNC_OUT to SYNC_IN for proper operation.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -689,10 +701,17 @@ static int ad7768_setup(struct ad7768_state *st)
 		return ret;
 
+	st->en_gpio_sync = false;
+	st->en_spi_sync = false;
```
Ok!

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -689,10 +701,17 @@ static int ad7768_setup(struct ad7768_state *st)
+	if (!IS_ERR(st->gpio_sync_in))
+		st->en_gpio_sync = true;
+
+	/* GPIO and SPI Synchronization are mutually exclusive */
```
indeed, we are not returning error when both are added.  i will fix this

On the current approach, we accept neither of them. in this case, we consider that a external synchronization pulse is provided.

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -689,10 +701,17 @@ static int ad7768_setup(struct ad7768_state *st)
-	if (IS_ERR(st->gpio_sync_in))
-		return PTR_ERR(st->gpio_sync_in);
+	if (!IS_ERR(st->gpio_sync_in))
+		st->en_gpio_sync = true;
```
You are right, that is better

jonathanns commented on drivers/iio/adc/ad7768-1.c:
```diff
@@ -180,6 +220,15 @@ static struct iio_chan_spec_ext_info ad7768_ext_info[] = {
+		.write = ad7768_ext_info_write,
+		.shared = IIO_SHARED_BY_ALL,
+		.private = DEC_RATE,
+	},
```
Ok, I will do that!

jonathanns commented:
v7:
* Replaced `mutex_lock()` and `mutex_unlock()` with `cleanup.h` tools.
* Moved multiple scan types commit; Filter and decimation attributes are added without bugs;
* Added a helper function to set decimation and filter, reducing code size.
* Fixed the AAF gain check: return an error if the gain is set but not supported, or if it is required but not defined for the part.
* Added a mutual exclusivity check for `syn-in-spi` and `syn-in-gpio` properties.
* Applied a hack for `rx_buf` from offload transfer, as suggested by Marcelo.
* Specified a 16-bit size for the `adi,aaf-gain` property.
* Addressed other nits raised by Nuno and Marcelo.

machschmitt reviewed (approved):
Looks like our managers are getting more open to the idea of going upstream first :man_dancing: . Our lives will get much better if we be able to work upstream and change only what can't be upstreamed (at the moment) in ADI Linux tree. That way we will save a lot of (re)work that is often required to develop (and maintain) drivers in both trees because code always change during the upstreaming process. Approving this as I think the code is in okayish shape to go to the mailing list (note you will have to drop some stuff related to offload support).
```bash
git clone git://git.kernel.org/pub/scm/linux/kernel/git/jic23/iio.git
cd iio
git fetch origin testing:testing
git checkout testing
```
Now the real work can begin :)

nunojsa commented:
Hi, I'm just checking in on the status of this pull request. Is there anything I can do to help move it forward?

nunojsa commented:
> Also, I’ve been discussing with some colleagues about how to handle backporting those patches, some are modified versions of existing patches in this tree (the upstream was way behind this one). Would you prefer that I wipe this driver and replace it with the upstream version, or create new patches adapting what we currently have

No strong opinion. Slight preference for wiping out what we have now (sync it with an upstream "base" version) and then cleanly apply your upstream patches. You just need to make sure that all important features are either upstreamed or if not possible for some reason, they are kept in the ADI tree.

I would anyways close this PR (if ok with you) and then open a new one when you're ready for it.
